R/derive_param_tte.R

```diff
@@ -0,0 +1,616 @@
+#' library(dplyr, warn.conflicts = FALSE)
+#' library(lubridate, warn.conflicts = FALSE)
+#' data("adsl")
+#'
+#' death <- tte_source(
+#'   dataset_name = "adsl",
+#'   filter = DTHFL == "Y",
+#'   date = DTHDT,
+#'   set_values_to =vars(
+#'     EVENTDESC = "DEATH",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "DTHDT"))
+#'
+#' lstalv <- tte_source(
+#'   dataset_name = "adsl",
+#'   date = LSTALVDT,
+#'   censor = 1,
+#'   set_values_to = vars(
+#'     EVENTDESC = "LAST KNOWN ALIVE DATE",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "LSTALVDT"))
+#'
+#' derive_param_tte(
+#'   dataset_adsl = adsl,
+#'   event_conditions = list(death),
+#'   censor_conditions = list(lstalv),
+#'   source_datasets = list(adsl = adsl),
+#'   set_values_to = vars(PARAMCD = "OS",
+#'                        PARAM = "Overall Survival")) %>%
+#' select(-STUDYID) %>% `[`(20:30,)
+#'
+#' # derive time to adverse event for each preferred term #
+#' adsl <- tibble::tribble(
+#' ~USUBJID, ~TRTSDT,           ~EOSDT,
+#' "01",     ymd("2020-12-06"), ymd("2021-03-06"),
+#' "02",     ymd("2021-01-16"), ymd("2021-02-03")) %>%
+#'   mutate(STUDYID = "AB42")
+#'
+#' ae <- tibble::tribble(
+#'   ~USUBJID, ~AESTDTC,           ~AESEQ, ~AEDECOD,
+#'   "01",     "2021-01-03T10:56", 1,      "Flu",
+#'   "01",     "2021-03-04",       2,      "Cough",
+#'   "01",     "2021",             3,      "Flu") %>%
+#'   mutate(STUDYID = "AB42")
+#'
+#' ttae <- tte_source(
+#'   dataset_name = "ae",
+#'   date = AESTDTC,
+#'   set_values_to = vars(
+#'     EVENTDESC = "AE",
+#'     SRCDOM = "AE",
+#'     SRCVAR = "AESTDTC",
+#'     SRCSEQ = AESEQ))
+#'
+#' eos <- tte_source(
+#'   dataset_name = "adsl",
+#'   date = EOSDT,
+#'   censor = 1,
+#'   set_values_to = vars(
+#'     EVENTDESC = "END OF STUDY",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "EOSDT"))
+#'
+#' derive_param_tte(
+#'   dataset_adsl = adsl,
+#'   by_vars = vars(AEDECOD),
+#'   start_date = TRTSDT,
+#'   event_conditions = list(ttae),
+#'   censor_conditions = list(eos),
+#'   source_datasets = list(adsl = adsl, ae = ae),
+#'   set_values_to = vars(
+#'     PARAMCD = paste0("TTAE", as.numeric(as.factor(AEDECOD))),
+#'     PARAM = paste("Time to First", AEDECOD, "Adverse Event"),
+#'     PARCAT1 = "TTAE",
+#'     PARCAT2 = AEDECOD
+#'   )
+#' ) %>%
+#' select(USUBJID, STARTDT, PARAMCD, PARAM, ADT, CNSR, SRCSEQ)
+derive_param_tte <- function(dataset = NULL,
+                             dataset_adsl,
+                             by_vars = NULL,
+                             start_date = TRTSDT,
+                             start_date_imputation_flag = NULL,
+                             start_time_imputation_flag = NULL,
+                             event_conditions,
+                             censor_conditions,
+                             source_datasets,
+                             create_datetime = FALSE,
+                             set_values_to,
+                             subject_keys = vars(STUDYID, USUBJID)) {
```

Comment:
Do we need a special parameter for ADSL or could we just make it a required element of `source_datasets`?. Anyway, I would move the `source_datasets` parameter much further to the top.